feat: Add more list operations

aborgna-q opened the pull request:
Closes #1467.

`find` requires #1473

zrho commented on hugr-core/src/std_extensions/collections.rs:
```diff
@@ -134,10 +151,19 @@ impl ListOp {
+            lookup => self.list_polytype(vec![l, USIZE_T], vec![e]).into(),
+            length => self.list_polytype(vec![l], vec![USIZE_T]).into(),
+            repeat => self.list_polytype(vec![l.clone(), USIZE_T], vec![l]).into(),
+            count => self.list_polytype(vec![l, e], vec![USIZE_T]).into(),
```
How are the elements compared when counted? We don't have a constraint system yet that could ensure that the elements have a well-defined equality operation. Until then we could make this a dynamic thing of course. Or we could pass in an equality predicate?

aborgna-q commented on hugr-core/src/std_extensions/collections.rs:
```diff
@@ -134,10 +151,19 @@ impl ListOp {
-            pop => self
-                .list_polytype(vec![l.clone()], vec![l.clone(), e.clone()])
-                .into(),
+            pop => self.list_polytype(vec![l.clone()], vec![l, e]).into(),
```
Now there's both! `pop`/`pop_unchecked` and `lookup`/`lookup_unchecked`.

I implemented #1473 as part of it, and added `find`.

aborgna-q commented:
I removed all panicking operations, as they can be defined from checked ones (with an if and a standalone panic).

I also reduced the set of operations to a core one. This gets rid of the equality checks problem.

I moved the option/result definitions to it's own PR (#1481). I'll rebase on top of it.

ss2165 commented:
regarding panicking, how about adding a generic `unwrap`: `forall R, E. [R, E] -> R` to the prelude?

aborgna-q commented:
It seems I'll actually need an unwrap to fix this test.
https://github.com/CQCL/hugr/blob/602fa86eacca38022349e1005203e1ee1d9494fe/hugr-passes/src/const_fold/test.rs#L126-L131
(It was skipped due to some inference error, but it works now).

I'd prefer to skip it temporarily, and add the unwrap in a separate PR.

ss2165 reviewed (approved):
:shipit:
